application/controllers/OptinController.php

```diff
@@ -27,53 +27,183 @@ class OptinController extends LSYii_Controller
+            throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
+        } else {
+            LimeExpressionManager::singleton()->loadTokenInformation($surveyId, $accessToken, false);
+            $token = Token::model($surveyId)->findByAttributes(['token' => $accessToken]);
+
+            if (!isset($token)) {
+                $message = gT('You are not a participant of this survey.');
+            } else {
+                if ($token->emailstatus != 'OptOut') {
+                    $message = gT('You are already a participant of this survey.');
+                } else {
+                    $message = "<p>" . gT('Please confirm that you want to be added back to this survey by clicking the button below.') . '<br>' . gT("After confirmation you may start receiving invitations and reminders for this survey.") . "</p>";
+                    $message .= '<p><a href="' . Yii::app()->createUrl('optin/addtokens', ['surveyid' => $surveyId, 'langcode' => $baseLanguage, 'token' => $accessToken]) . '" class="btn btn-default btn-lg">' . gT("I confirm") . '</a><p>';
+                }
+            }
+        }
+
+        $this->renderHtml($message, $survey);
+    }
+
+    /**
+     * Display the confirmation for global opt in
+     */
+    public function actionparticipants()
+    {
+        Yii::app()->loadHelper('database');
+        Yii::app()->loadHelper('sanitize');
+
+        $surveyId = Yii::app()->request->getQuery('surveyid');
+        $languageCode = Yii::app()->request->getQuery('langcode');
+        $accessToken = Token::sanitizeToken(Yii::app()->request->getQuery('token'));
+
+        //IF there is no survey id, redirect back to the default public page
+        if (!$surveyId) {
+            $this->redirect(['/']);
+        }
+
+        $surveyId = (int) $surveyId; //Make sure it's an integer (protect from SQL injects)
+        $survey = Survey::model()->findByPk($surveyId);
+
+        //Check that there is a SID
+        // Get passed language from form, so that we dont lose this!
+        if (!isset($languageCode) || $languageCode == "" || !$languageCode) {
+            $baseLanguage = $survey->language;
+        } else {
+            $baseLanguage = sanitize_languagecode($languageCode);
+        }
+
+        Yii::app()->setLanguage($baseLanguage);
+
+        if (empty($survey) || !$survey->hasTokensTable) {
+            throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
+        } else {
+            LimeExpressionManager::singleton()->loadTokenInformation($surveyId, $accessToken, false);
+            $token = Token::model($surveyId)->findByAttributes(['token' => $accessToken]);
+
+            if (!isset($token)) {
+                $message = "<p>" . gT('You are not a participant of this survey.') . "</p>";
```

Comment:
tag in PHP : must be theme twig file

Reply:
See https://github.com/LimeSurvey/LimeSurvey/blob/master/themes/survey/vanilla/views/subviews/registration/register_message.twig for example

Reply:
Not related, but let me tell you something I don't love that much about this twig.
Twig idea is for sperating controller/logic and view.
But in that twig the status is kind of decided on the view and not the controller/logic.
(just as some feedback, not to make a discussion out of it) :)

Reply:
It's currently the only solution to allow user to do such things : https://forums.limesurvey.org/forum/can-i-do-this-with-limesurvey/126630-access-code-page#224661 

I really think it's something needed and great for user.

Reply:
> allow user to do such things

You mean "Override messages from theme"?

Reply:
For example ,
Or rediect in some condition or use different color or … etc …

Reply:
>  But in that twig the status is kind of decided on the view and not the controller/logic.

Showin specific string is view task : not logic task. But create a twig file for each message seems overkiil

Reply:
Redirect doesn't sound like a "view" task.

> Showin specific string is view task

Display message is view task. 
Deciding which message to display is "view" task also?
I understand we have to give in on purity sometimes. Not complaining. Just sending feedback :)

Reply:
>  Deciding which message to display is "view" task also?

A possible «««good»»» way is to load a specific twig for each message type. But create a twig file for each message seems overkiil :)

> Redirect doesn't sound like a "view" task.

Yeag, right, but user like it i think (see the number of forum post about way to redirect Survey list page for example).

>  Deciding which message to display is "view" task also?

No : but have "error type" and choose message according to type ? Error information/detail .

When try to construct real solution by theme : it's really hard to get information. Currently sometimes i use ExpressionManager to get core information (for example {{ processString("{TOKEN}") }}seems really a workaround and not a good way to do.
If we are inside Participant management (for example), seems legit to have Participant information.

Reply:
Ok, i report the issue tomorrow :).

Close this one